tests/5.1/dispatch/test_dispatch_is_device_ptr.c

```diff
@@ -0,0 +1,74 @@
+//===---- test_dispatch_is_device_ptr.c -----------------------------------------===//
+// 
+// OpenMP API Version 5.1
+//
+// Uses dispatch construct as context for variant directive. Uses is_device_ptr
+// clause to ensure that the list item is in the device region.
+//
+// Inspired by "OpenMP 5.1 Features: The Dispatch Construct" video:
+// https://www.youtube.com/watch?v=ruugaX95gIs
+// 
+//===-------------------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include "ompvv.h"
+#include <stdbool.h>
+
+#define N 1024
+
+int errors;
+int i = 0;
+int *arr;
+
+void add_dev(int *arr);
+
+#pragma omp declare variant(add_dev) match(construct={dispatch}) 
+void add(int *arr){
+    #pragma omp parallel for
+    for (int i = 0; i < N; i++){ // Base function adds 2 to array values
+        arr[i] = arr[i]+2;
+    }
+}
+
+void add_dev(int *arr){
+    #pragma omp target is_device_ptr(arr)
+    for (int i = 0; i < N; i++){
+        arr[i] = arr[i]+4; // Variant function adds 4 to array values
+    }
+}
+
+int test_wrapper() { 
+    int t;
+    errors = 0;
+    t = omp_get_default_device();
+    arr = (int *)omp_target_alloc( sizeof(int)*N, t);
+    if(arr == NULL){
+        OMPVV_ERROR("Can't properly create device pointer");
+        return 1;
+    }
+    #pragma omp target is_device_ptr(arr)
+    {
+        #pragma omp parallel for
+        for(int i = 0; i < N; i++){
+            arr[i] = i;
+        }
+    }
+    #pragma omp dispatch is_device_ptr(arr)
+        add(arr);
+    
+    #pragma omp target map(tofrom: errors)
+    for(i = 0; i < N; i++){
+        OMPVV_TEST_AND_SET(errors, (arr[i] != i+4) || (arr[i] != i+2) );
```

Comment:
Here we can add a INFOMSG here if (arr[i] == i+2) that the dispatch construct in the context-selector was ignored.

Reply:
BTW: It would be good if the INFOMSG is only printed once and not for every of the N=1024 array elements.

Regarding the code: If 'add' is called, i.e. the base function: That's only valid if omp_get_default_device() is the initial device **or** there is unified shared memory.  – Because 'add' is called on the host and 'arr' is allocated on the device with 'omp_target_alloc(...)'.

IMHO: I think that the code requires a comment why both results are valid. In my understanding, it is an oversight in the spec that it is implementation defined which one is invoked – fixed in 6.0, isn't it?
In any case, I think a comment is required here – and in the other files – stating why "== i +2" is okay.

Reply:
Should `||` be changed to `&&`?

Reply:
@tob2 , yes, you  are correct. This behavior is allowed in 5.1 but will change in 6.0.